sys/arm64/include/cherireg.h

```diff
@@ -0,0 +1,167 @@
+* executable capabilities (pcc); CHERI_PERM_STORE, CHERI_PERM_STORE_CAP,
+ * and CHERI_PERM_STORE_LOCAL_CAP will be added for data permissions (ddc).
+ */
+#define	CHERI_PERMS_USERSPACE						\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_LOAD | CHERI_PERM_LOAD_CAP |	\
+	(CHERI_PERMS_SWALL & ~CHERI_PERM_CHERIABI_VMMAP))
+
+#define	CHERI_PERMS_USERSPACE_CODE					\
+	(CHERI_PERMS_USERSPACE | CHERI_PERM_EXECUTE)
+
+#define	CHERI_PERMS_USERSPACE_SEALCAP					\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_SEAL | CHERI_PERM_UNSEAL)
+
+/*
+ * _DATA includes _VMMAP so that we can derive the MMAP cap from it.
+ *
+ * XXX: We may not want to include VMMAP here and instead only in
+ * CHERI_CAP_USER_MMAP_PERMS
+ */
+#define	CHERI_PERMS_USERSPACE_DATA					\
+				(CHERI_PERMS_USERSPACE |		\
+				CHERI_PERM_STORE |			\
+				CHERI_PERM_STORE_CAP |			\
+				CHERI_PERM_STORE_LOCAL_CAP |		\
+				CHERI_PERM_CHERIABI_VMMAP)
+
+/*
+ * Corresponding permission masks for kernel code and data; these are
+ * currently a bit broad, and should be narrowed over time as the kernel
+ * becomes more capability-aware.
+ */
+#define	CHERI_PERMS_KERNEL						\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_LOAD | CHERI_PERM_LOAD_CAP)	\
+
+#define	CHERI_PERMS_KERNEL_CODE						\
+	(CHERI_PERMS_KERNEL | CHERI_PERM_EXECUTE | CHERI_PERM_SYSTEM_REGS)
+
+#define	CHERI_PERMS_KERNEL_DATA						\
+	(CHERI_PERMS_KERNEL | CHERI_PERM_STORE | CHERI_PERM_STORE_CAP |	\
+	CHERI_PERM_STORE_LOCAL_CAP)
+
+#define	CHERI_PERMS_KERNEL_SEALCAP					\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_SEAL | CHERI_PERM_UNSEAL)
+
+/*
+ * The CHERI object-type space is split between userspace and kernel,
+ * permitting kernel object references to be delegated to userspace (if
+ * desired).  Currently, we provide 13 bits of namespace to each, with the top
+ * bit set for kernel object types, but it is easy to imagine other splits.
+ * User and kernel software should be written so as to not place assumptions
+ * about the specific values used here, as they may change.
+ */
+#define	CHERI_OTYPE_BITS	(14)
+#define	CHERI_OTYPE_USER_MIN	(0)
+#define	CHERI_OTYPE_USER_MAX	((1 << (CHERI_OTYPE_BITS - 1)) - 1)
+#define	CHERI_OTYPE_KERN_MIN	(1 << (CHERI_OTYPE_BITS - 1))
+#define	CHERI_OTYPE_KERN_MAX	((1 << CHERI_OTYPE_BITS) - 1)
+#define	CHERI_OTYPE_KERN_FLAG	(1 << (CHERI_OTYPE_BITS - 1))
+#define	CHERI_OTYPE_ISKERN(x)	(((x) & CHERI_OTYPE_KERN_FLAG) != 0)
+#define	CHERI_OTYPE_ISUSER(x)	(!(CHERI_OTYPE_ISKERN(x)))
+
+/*
+ * When performing a userspace-to-userspace CCall, capability flow-control
+ * checks normally prevent local capabilities from being delegated.  This can
+ * be disabled on call (but not return) by using an object type with the 22nd
+ * bit set -- combined with a suitable selector on the CCall instruction to
+ * ensure that this behaviour is intended.
+ */
+#define	CHERI_OTYPE_LOCALOK_SHIFT	(CHERI_OTYPE_BITS - 2)
+#define	CHERI_OTYPE_LOCALOK_FLAG	(1 << CHERI_OTYPE_LOCALOK_SHIFT
+#define	CHERI_OTYPE_IS_LOCALOK(x)	(((x) & CHERI_OTYPE_LOCALOK_FLAG) != 0)
```

Comment:
This is all MIPS-specific (well, the comment at least) and I assume dates back to the old exception-based CCall/CReturn mechanism. RISC-V doesn't have any of this.